File: app/src/main/java/co/livil/nna/androidautodemo/NotificationHandler.java
package co.livil.nna.androidautodemo;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.core.app.NotificationCompat;
import androidx.core.app.Person;
import androidx.core.app.RemoteInput;

import org.jetbrains.annotations.NotNull;

import co.livil.nna.androidautodemo.R.drawable;
import co.livil.nna.androidautodemo.R.string;
import kotlinx.coroutines.flow.FlowKt;
import kotlinx.coroutines.flow.MutableStateFlow;
import kotlinx.coroutines.flow.StateFlow;
import kotlinx.coroutines.flow.StateFlowKt;

public final class NotificationHandler {
    private int notificationCount;
    private final Context context;
    @NotNull
    public static final String CHANNEL_ID = "nna_channel_id";
    @NotNull
    public static final String ACTION_REPLY = "co.livil.nna.androidautodemo.NOTIFICATION_ACTION_REPLY";
    @NotNull
    public static final String ACTION_MARK_AS_READ = "co.livil.nna.androidautodemo.NOTIFICATION_ACTION__MARK_AS_READ";
    @NotNull
    public static final String KEY_TEXT_REPLY = "key_text_reply";
    @NotNull
    public static final String EXTRA_NOTIFICATION_ID_KEY = "extra_notification_id_key";
    private static final MutableStateFlow _replies = StateFlowKt.MutableStateFlow((Object)null);
    @NotNull
    public static final StateFlow replies;

    private final int getCurrentNotificationId() {
        return 1000 + this.notificationCount;
    }

    public final void createNotificationChannel() {
        String channelId = "nna_channel_id";
        String channelName = this.context.getString(string.notification__channel_name);
        NotificationChannel channel = new NotificationChannel(channelId, (CharSequence)channelName, NotificationManager.IMPORTANCE_DEFAULT);

        channel.setDescription(this.context.getString(string.notification__channel_description));
        Object service = this.context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (service == null) {
            throw new NullPointerException("null cannot be cast to non-null type android.app.NotificationManager");
        } else {
            NotificationManager notificationManager = (NotificationManager)service;
            notificationManager.createNotificationChannel(channel);
        }
    }

    public final void sendNotification() {
        Object service = this.context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (service == null) {
            throw new NullPointerException("null cannot be cast to non-null type android.app.NotificationManager");
        } else {
            NotificationManager notificationManager = (NotificationManager)service;
            Notification notification = this.createMessageNotification();
            notificationManager.notify(this.getCurrentNotificationId(), notification);
        }
    }

    private final Notification createMessageNotification() {
        String channelId = "nna_channel_id";
        this.notificationCount++;

        Person sender = (new Person.Builder()).setName((CharSequence)"Sender").build();
        NotificationCompat.MessagingStyle style = (new NotificationCompat.MessagingStyle(sender))
                .addMessage((CharSequence)("New message " + this.notificationCount), System.currentTimeMillis(), sender);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(this.context, channelId);
        builder.setSmallIcon(drawable.ic_notification_icon);
        builder.setPriority(1);
        builder.setAutoCancel(true);
        builder.addAction(this.createReplyAction());
        builder.addAction(this.createMarkAsReadAction());
        builder.setStyle((NotificationCompat.Style)style);

        return builder.build();
    }

    private final NotificationCompat.Action createReplyAction() {
        Intent replyIntent = new Intent(this.context, NotificationActionReceiver.class);
        replyIntent.setAction("co.livil.nna.androidautodemo.NOTIFICATION_ACTION_REPLY");
        replyIntent.putExtra("extra_notification_id_key", this.getCurrentNotificationId());

        int replyFlags = PendingIntent.FLAG_UPDATE_CURRENT;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            replyFlags = PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_MUTABLE;
        }

        PendingIntent replyPendingIntent = PendingIntent.getBroadcast(this.context, 0, replyIntent, replyFlags);
        RemoteInput.Builder builder = new RemoteInput.Builder("key_text_reply");
        builder.setLabel((CharSequence)"Reply");
        RemoteInput remoteInput = builder.build();

        NotificationCompat.Action remoteAction = (new NotificationCompat.Action.Builder(drawable.ic_reply_icon, (CharSequence)"Reply", replyPendingIntent))
                .setSemanticAction(NotificationCompat.Action.SEMANTIC_ACTION_REPLY)
                .setShowsUserInterface(false)
                .addRemoteInput(remoteInput)
                .build();

        return remoteAction;
    }

    private final NotificationCompat.Action createMarkAsReadAction() {
        Intent intent = new Intent(this.context, NotificationActionReceiver.class);
        intent.setAction("co.livil.nna.androidautodemo.NOTIFICATION_ACTION__MARK_AS_READ");
        intent.putExtra("extra_notification_id_key", this.getCurrentNotificationId());

        PendingIntent readPendingIntent = PendingIntent.getBroadcast(this.context, 1, intent, PendingIntent.FLAG_IMMUTABLE);
        NotificationCompat.Action remoteAction = (new NotificationCompat.Action.Builder(drawable.ic_read_icon, (CharSequence)"Mark as Read", readPendingIntent))
                .setSemanticAction(NotificationCompat.Action.SEMANTIC_ACTION_MARK_AS_READ)
                .setShowsUserInterface(false)
                .build();

        return remoteAction;
    }

    public NotificationHandler(@NotNull Context context) {
        super();
        this.context = context;
    }

    static {
        replies = FlowKt.asStateFlow(_replies);
    }

    @NotNull
    public final StateFlow getReplies() {
        return NotificationHandler.replies;
    }

    public static final void postReply(@NotNull String reply) {
        NotificationHandler._replies.setValue(reply);
    }
}

File: app/src/main/java/co/livil/nna/androidautodemo/NotificationHandler.kt
package co.livil.nna.androidautodemo

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.Person
import androidx.core.app.RemoteInput
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow

class NotificationHandler(private val context: Context) {
    companion object {
        const val CHANNEL_ID = "nna_channel_id"

        const val ACTION_REPLY = "co.livil.nna.androidautodemo.NOTIFICATION_ACTION_REPLY"
        const val ACTION_MARK_AS_READ = "co.livil.nna.androidautodemo.NOTIFICATION_ACTION__MARK_AS_READ"
        const val KEY_TEXT_REPLY = "key_text_reply"
        const val EXTRA_NOTIFICATION_ID_KEY = "extra_notification_id_key"

        private val _replies: MutableStateFlow<String?> = MutableStateFlow(null)
        val replies = _replies.asStateFlow()

        fun postReply(reply: String) {
            _replies.value = reply
        }
    }

    private var notificationCount = 0
    private val currentNotificationId
        get() = 1000 + notificationCount

    fun createNotificationChannel() {
        val channelId = CHANNEL_ID
        val channelName = context.getString(R.string.notification__channel_name)
        val importance = NotificationManager.IMPORTANCE_DEFAULT
        val channel = NotificationChannel(channelId, channelName, importance).apply {
            description = context.getString(R.string.notification__channel_description)
        }
        // Register the channel with the system
        val notificationManager: NotificationManager =
            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.createNotificationChannel(channel)
    }

    fun sendNotification() {
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val notification = createMessageNotification() // The method you created earlier
        notificationManager.notify(currentNotificationId, notification)
    }

    private fun createMessageNotification(): Notification {
        val channelId = CHANNEL_ID

        notificationCount++

        val sender = Person.Builder().setName("Sender").build()
        val style = NotificationCompat.MessagingStyle(sender)
            .addMessage("New message $notificationCount", System.currentTimeMillis(), sender)

        // Create the Notification
        val builder = NotificationCompat.Builder(context, channelId).apply {
            setSmallIcon(R.drawable.ic_notification_icon)
            //setContentTitle("New Message ${notificationCount}")
            //setContentText("You've received a new message.")
            priority = NotificationCompat.PRIORITY_HIGH
            setAutoCancel(true)

            // Both of these actions are required for Android Auto to display the notification
            addAction(createReplyAction())
            addAction(createMarkAsReadAction())

            setStyle(style)
        }

        return builder.build()
    }

    private fun createReplyAction(): NotificationCompat.Action {
        // Intent for the reply action
        val replyIntent = Intent(context, NotificationActionReceiver::class.java).apply {
            action = ACTION_REPLY
            putExtra(EXTRA_NOTIFICATION_ID_KEY, currentNotificationId)
        }
        val replyFlags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
        } else {
            PendingIntent.FLAG_UPDATE_CURRENT
        }

        val replyPendingIntent: PendingIntent = PendingIntent.getBroadcast(context, 0, replyIntent, replyFlags)

        // RemoteInput for the reply action
        val remoteInput: RemoteInput = RemoteInput.Builder(KEY_TEXT_REPLY).run {
            setLabel("Reply")
            build()
        }

        // Reply action
        return NotificationCompat.Action.Builder(R.drawable.ic_reply_icon, "Reply", replyPendingIntent)
            // Provide context to what firing Action does.
            .setSemanticAction(NotificationCompat.Action.SEMANTIC_ACTION_REPLY)
            // The action doesn't show any UI (it's a requirement for apps to not show UI for Android Auto).
            .setShowsUserInterface(false)
            // Add remote input to the action
            .addRemoteInput(remoteInput)
            .build()
    }

    private fun createMarkAsReadAction(): NotificationCompat.Action {
        // PendingIntent for mark as read action
        val readIntent = Intent(context, NotificationActionReceiver::class.java).apply {
            action = ACTION_MARK_AS_READ
            putExtra(EXTRA_NOTIFICATION_ID_KEY, currentNotificationId)
        }
        val readPendingIntent: PendingIntent = PendingIntent.getBroadcast(
            context,
            1,
            readIntent,
            PendingIntent.FLAG_IMMUTABLE
        )

        return NotificationCompat.Action.Builder(R.drawable.ic_read_icon, "Mark as Read", readPendingIntent)
            // Provide context to what firing Action does.
            .setSemanticAction(NotificationCompat.Action.SEMANTIC_ACTION_MARK_AS_READ)
            // The action doesn't show any UI (it's a requirement for apps to not show UI for Android Auto).
            .setShowsUserInterface(false)
            .build()
    }

}